API/hermes_abi/HermesABIRuntime.cpp

```diff
@@ -0,0 +1,904 @@
+    }
+
+    /// Set the next free element. Must not be called when this instance is
+    /// occupied with a value.
+    void setNextFree(ManagedPointerHolder *nextFree) {
+      assert(isFree() && "Free pointer unusuable while occupied");
+      nextFree_ = nextFree;
+    }
+
+    HermesABIManagedPointer *getManagedPointer() const {
+      assert(!isFree() && "Value not present");
+      return managedPointer_;
+    }
+
+    void invalidate() override {
+      dec();
+    }
+
+    void inc() {
+      // See comments in hermes_abi.cpp for why we use relaxed operations here.
+      auto oldCount = refCount_.fetch_add(1, std::memory_order_relaxed);
+      assert(oldCount && "Cannot resurrect a pointer");
+      assert(oldCount + 1 != 0 && "Ref count overflow");
+      (void)oldCount;
+    }
+
+    void dec() {
+      // See comments in hermes_abi.cpp for why we use relaxed operations here.
+      auto oldCount = refCount_.fetch_sub(1, std::memory_order_relaxed);
+      assert(oldCount > 0 && "Ref count underflow");
+      // This was the last decrement of this holder, so we can invalidate the
+      // underlying pointer.
+      if (oldCount == 1)
+        managedPointer_->vtable->invalidate(managedPointer_);
+    }
+
+   private:
+    std::atomic<uint32_t> refCount_;
+    union {
+      HermesABIManagedPointer *managedPointer_;
+      ManagedPointerHolder *nextFree_;
+    };
+  };
+
+  /// An implementation of HermesABIByteBuffer that uses a string as its
+  /// internal storage. This can be used to conveniently construct a std::string
+  /// from ABI functions that return strings.
+  class StringByteBuffer : public HermesABIByteBuffer {
+    std::string buf_;
+
+    static void grow_by(HermesABIByteBuffer *buf, size_t amount) {
+      auto *self = static_cast<StringByteBuffer *>(buf);
+      self->buf_.resize(self->buf_.size() + amount);
+      self->data = (uint8_t *)self->buf_.data();
+      self->available += amount;
+    }
+
+    static constexpr HermesABIByteBufferVTable vt{
+        grow_by,
+    };
+
+   public:
+    explicit StringByteBuffer() : HermesABIByteBuffer{&vt, nullptr, 0} {
+      // Make the small string storage available for use without needing to call
+      // grow_by.
+      buf_.resize(buf_.capacity());
+      data = (uint8_t *)buf_.data();
+      available = buf_.size();
+    }
+
+    std::string get() && {
+      // Trim off any unused bytes at the end.
+      buf_.resize(buf_.size() - available);
+      return std::move(buf_);
+    }
+  };
+
+  class HostFunctionWrapper : public HermesABIHostFunction {
+    HermesABIRuntime &rt_;
+    HostFunctionType hf_;
+
+    static HermesABIValueOrError call(
+        HermesABIHostFunction *hf,
+        HermesABIContext *ctx,
+        const HermesABIValue *thisArg,
+        const HermesABIValue *args,
+        size_t count) {
+      auto *self = static_cast<HostFunctionWrapper *>(hf);
+      auto &rt = self->rt_;
+      try {
+        std::vector<Value> jsiArgs;
```

Comment:
It is better to use `SmallVector` here. The `std::vector` always allocates heap memory.

Reply:
The set of Hermes source used here is intentionally restricted. It is likely that integrators will want to build this C++ wrapper outside the main Hermes build.

Reply:
Yeah, you are right. I did not realize that it is part of the API for the user code.
Then, maybe we can have a hybrid strategy here: use a fixed array until a specified limit and only after that allocate on heap.
The issue is that this code is called for each function call and should avoid the memory allocation if we can.

Reply:
It is worth keeping in mind that once the representations of ABI values and JSI values are the same, we won't need any intermediate storage here at all.